src/cr/cube/slices.py

```diff
@@ -10,6 +10,200 @@
 from cr.cube.enum import DIMENSION_TYPE as DT
 
 
+class FrozenSlice(object):
+    """Main point of interaction with the outer world."""
+
+    def __init__(
+        self,
+        cube,
+        slice_idx=0,
+        use_insertions=False,
+        reordered_ids=None,
+        pruning=False,
+        weighted=True,
+        population=None,
+        ca_as_0th=None,
+    ):
+        self._cube = cube
+        self._slice_idx = slice_idx
+        self._use_insertions = use_insertions
+        self._reordered_ids = reordered_ids
+        self._pruning = pruning
+        self._weighted = weighted
+        self._population = population
+        self._ca_as_0th = ca_as_0th
+
+    # interface ----------------------------------------------------------------------
+
+    @lazyproperty
+    def base_counts(self):
+        return self._calculator.base_counts
+
+    @lazyproperty
+    def column_base(self):
```

Comment:
One thing that jumped out at me after I ordered these alphabetically is the pattern between column, row, and table. Column and row both have base, labels, margin, percentages, and proportions. Table has most of those (no labels). For now it's just an interesting symmetry to me, but makes me suspect an extractable object somewhere. I'm not sure what would be extracted though since these are all one-liners :). Anyway, too early for me to say there's anything to do about it, but an interesting pattern :) I really like that symmetry in the interface, that will make it much simpler to learn, because once you have the row, column, and table concepts, and the base, labels, margin, percentages, and proportions concepts, you basically understand the whole thing. Eight concepts to an interface? That's as simple as they come :)

Reply:
Yes! I've noticed that as well! I'm not sure how to deal with it. A little background: It didn't originate like that at all (a vector was responsible for all, nothing was on the slice). But with the introduction of the `table_margin`, and the decision that it be passed to the vector at construction time, we ended up in these shenanigans. So there might be a couple of things to tackle this:

1. At vector construction time (either row or column) - just pass it the whole slice, and the correct indexing (for the counts and bases) structure. It would go something like this:
`_CategoricalVector(self._row_dimension, index, self) for index in self._row_indexing`, or the like (not sure if we need row dimension in there, but the vector will need to know (maybe) if it's a row or a col, again, maybe in order to calculate something.
2. We might limit the `table_percentages` or `zscores` to only rows or only columns, so we just propagate the symmetrical methods to them. In this class we'd differentiate between a row and a col, which we now don't.

But I do like the concept of a particular vector not having to care about whether it's a row or a col. I think it removes a lot of responsibility from it, and consequently from duplicating the logic for that responsibility in two places (rows and columns).

Reply:
Yes, the bit about a vector not having to care whether it's a row or column sounds right to me. We could work at the next-higher abstraction level where it perhaps has an "opposing" dimension (row if its a col vector, col-vector if its rows). I especially think this makes sense because a "table" should be pivotable without any change necessary to the logic. An interface property like `.row_percentages` would still be on the slice (because that's how the user thinks about it), but it's implementation might be something like `return self._row_vectors.percentages` or something.

I think the take-away for me here is that rows and columns is a "user-interface" concept (the user here being the developer consuming the FrozenSlice interface properties) and should _not_ have technical/object coupling unless we discover it absolutely has to.

So for example, maybe there's a Vectors object and it is a collection of vectors and kind of also represents a dimension. Maybe it has a type, like MR or CAT, but definitely it at least contains vectors of a type, like MR-subvar vectors or CAT vectors or whatever. But mostly it's just the collection. Then the slice has a `._row_vectors` item that's a Vectors object and it has a `._column_vectors` item that is a Vectors object. Maybe some vectors objects also get their "opposing" vectors object (which makes for an interesting chicken-and-egg construction problem, but deferring that discussion for now). Then MR x CAT uses the same vectors as CAT x MR, just in a different position.

Anyway, just food for thought at this point. I'd say if we're within range of making it work, that's the first objective, because refactoring based on a fully working set of integration tests is a much more orderly environment and allows incremental refinement.

Reply:
Another possibility for situations where a Vectors object needs its opposing vectors for a calculation is to make that a method that takes the opposing vectors object. So one example, maybe not a good one, might be `slice.table_percentage` which could be implemented something like `self._row_vectors.table_percentage(self._column_vectors)`.

Because `Slice.table_percentage` is a lazyproperty, it would never calculate more than once. And if table-percentages are never needed locally, like as a basis for some other Vectors calculation, then the method would be called at most once for a Slice object.

Anyway, something to consider. Methods are not necessarily evil, they just have a cost we like to avoid whenever we can :)